### ISSUE TITLE
Mono debugger crashes in native code when inspecting interface instances produced by Config.Net

### ISSUE THREAD

**Issue by Nihlus** [mono/mono#6577]
## Steps to Reproduce

1. Download the attached reproduction project
2. Restore nuget packages and compile
3. Run the project
4. Observe the expected output
5. Debug the project 
6. Inspect the local variable in the main method.
7. Observe the runtime crash.

## Current Behavior
At present, when inspecting variables containing an instance of a proxy interface created by Config.Net, the Mono debugger crashes in native code. If the variable is not inspected, the program behaves as expected and prints the stored value.

## Expected Behavior
The program should run as expected, and the Mono debugger should be able to inspect the variables without crashing.

### On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:
```
Mono JIT compiler version 5.4.1.6 (tarball Wed Nov  8 20:35:02 UTC 2017)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           __thread
	SIGSEGV:       altstack
	Notifications: epoll
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          supported, not enabled.
	GC:            sgen (concurrent by default)
```

### Stacktrace
https://gist.github.com/Nihlus/d3274a9cce07728c3b36ce2da8082492

### Attachments
[CrashRepro.zip](https://github.com/mono/mono/files/1643632/CrashRepro.zip)
https://github.com/aloneguid/config
Issue on the Config.Net repository: https://github.com/aloneguid/config/issues/49

**Comment by vargaz** [comment]
Can't reproduce with an upcoming version:
Mono JIT compiler version 5.8.0.110 (2017-10/c7ec6a03ae3 Mon Jan 15 13:55:23 EST 2018)

**Comment by Nihlus** [comment]
Confirmed, issue fixed in Mono 5.8 .
